openvasd/openvasd.c

```diff
@@ -0,0 +1,2251 @@
+
+  if (val == NULL)
+    return OPENVASD_INVALID_VALUE;
+
+  switch (opt)
+    {
+    case OPENVASD_CA_CERT:
+      (*conn)->ca_cert = g_strdup ((char *) val);
+      break;
+    case OPENVASD_CERT:
+      (*conn)->cert = g_strdup ((char *) val);
+      break;
+    case OPENVASD_KEY:
+      (*conn)->key = g_strdup ((char *) val);
+      break;
+    case OPENVASD_API_KEY:
+      (*conn)->apikey = g_strdup ((char *) val);
+      break;
+    case OPENVASD_SERVER:
+      (*conn)->server = g_strdup ((char *) val);
+      break;
+    case OPENVASD_HOST:
+      (*conn)->host = g_strdup ((char *) val);
+      break;
+    case OPENVASD_SCAN_ID:
+      (*conn)->scan_id = g_strdup ((const char *) val);
+      break;
+    case OPENVASD_PORT:
+    default:
+      (*conn)->port = *((int *) val);
+      break;
+    };
+
+  return OPENVASD_OK;
+}
+
+/** @brief Build a openvasd connector
+ *  @Description Recieve option name and value to build the openvasd connector
+ *
+ *  @param[in/out] conn   struct holding the openvasd connector information
+ *  @param[in] opt    option to set
+ *  @param[in] val    value to set
+ *
+ *  @return Return OK on success, otherwise error;
+ */
+openvasd_error_t
+openvasd_connector_free (openvasd_connector_t *conn)
+{
+  if (*conn == NULL)
+    return OPENVASD_OK;
+
+  g_free ((*conn)->ca_cert);
+  g_free ((*conn)->cert);
+  g_free ((*conn)->key);
+  g_free ((*conn)->server);
+  g_free ((*conn)->host);
+  g_free ((*conn)->scan_id);
+  g_free (*conn);
+  *conn = NULL;
+
+  return OPENVASD_OK;
+}
+
+/**
+ * @brief Free an openvasd response struct
+ *
+ * @param resp Response to be free()'d
+ */
+void
+openvasd_response_free (openvasd_resp_t resp)
+{
+  if (resp == NULL)
+    return;
+
+  g_free (resp->body);
+  g_free (resp->header);
+  g_free (resp);
+  resp = NULL;
+}
+
+/** @brief Initialize the string struct to hold the response
+ *
+ *  @param s[in/out] The string struct to be initialized
+ */
+void
+init_openvasd_stringstream (openvasd_stringstream *s)
+{
+  s->len = 0;
+  s->ptr = g_malloc0 (s->len + 1);
+  if (s->ptr == NULL)
+    {
+      g_warning ("%s: Error allocating memory for response", __func__);
+      return;
+    }
+  s->ptr[0] = '\0';
+}
+
+/** @brief Reinitialize the string struct to hold the response
+ *
+ *  @param s[in/out] The string struct to be initialized
+ */
+static void
+reset_openvasd_stringstream (openvasd_stringstream *s)
+{
+  g_free (s->ptr);
+  init_openvasd_stringstream (s);
+}
+
+/** @brief Call back function to stored the response.
+ *
+ *  @description The function signature is the necessary to work with
+ *  libcurl. It stores the response in s. It reallocate memory if necessary.
+ */
+static size_t
+response_callback_fn (void *ptr, size_t size, size_t nmemb, void *struct_string)
+{
+  openvasd_stringstream *s = struct_string;
+  size_t new_len = s->len + size * nmemb;
+  char *ptr_aux = g_realloc (s->ptr, new_len + 1);
+  s->ptr = ptr_aux;
+  if (s->ptr == NULL)
```

Comment:
Similar to `init_openvasd_stringstream`, g_realloc will abort when out of mem.

But I'm wondering why you're using your own stringstream structure instead of GString. To me it looks like the behaviour will be the same. Glib offers g_string_append_len to append with a certain size. Feels safer to me, and glib will probably also do the reallocs better (some sort of doubling algorithm instead of reallocating exactly what is needed every time).